oidc/oidc.go

```diff
@@ -24,28 +26,71 @@ type OIDC struct {
 	config
 	s        *securecookie.SecureCookie
 	loginURL string
+	Loader   *loader
+}
+
+type loader struct {
+	InitChan      chan struct{}
+	IsInitialized atomic.Bool
+	InitOnce      sync.Once
+	ErrChan       chan error
 }
 
 // New returns a new OIDC Authenticator
-func New(ctx context.Context, s *securecookie.SecureCookie, issuerURL, clientID, clientSecret, redirectURL string) (*OIDC, error) {
-	provider, err := oidc.NewProvider(ctx, issuerURL)
-	if err != nil {
-		return nil, errors.Wrap(err, "oidc.NewProvider()")
+func New(ctx context.Context, s *securecookie.SecureCookie, issuerURL, clientID, clientSecret, redirectURL string) *OIDC {
+	errChan := make(chan error) // TODO: find a time and place to consume error from errChan and cancel the context in Main()
+	initChan := make(chan struct{})
+
+	// return partial authenticator struct with sync primitives to safely return OIDC config
+	// after its finished initializing
+	authenticator := &OIDC{
+		s: s,
+		Loader: &loader{
+			InitChan:      initChan,
+			IsInitialized: atomic.Bool{},
+			InitOnce:      sync.Once{},
+			ErrChan:       errChan,
+		},
 	}
+	authenticator.Loader.IsInitialized.Store(false)
+
+	go func() {
+		defer close(initChan)
+		defer close(errChan)
+
+		provider, err := oidc.NewProvider(ctx, issuerURL)
+		// TODO: check err for delay on cold start, backoff & retry.
+		// otherwise send error to channel.
+		if err != nil {
+			errChan <- errors.Wrap(err, "oidc.NewProvider()")
+
+			return
+		}
 
-	return &OIDC{
-		provider: provider,
-		config: &oAuth2{
+		authenticator.provider = provider
+		authenticator.config = &oAuth2{
 			config: oauth2.Config{
 				ClientID:     clientID,
 				ClientSecret: clientSecret,
 				RedirectURL:  redirectURL,
 				Endpoint:     provider.Endpoint(),
 				Scopes:       []string{oidc.ScopeOpenID, "profile"},
 			},
-		},
-		s: s,
-	}, nil
+		}
+		authenticator.Loader.IsInitialized.Store(true)
+	}()
+
+	return authenticator
+}
+
+func (o *OIDC) IsAuthenticatorInitialized() bool {
+	o.Loader.InitOnce.Do(func() {
+		if !o.Loader.IsInitialized.Load() {
+			<-o.Loader.InitChan
```

Comment:
This read on channel `InitChan` will hang indefinently until the go-routine completes. This will block all usage. Usage should fail fast if we can not initiaized.